web/apps/labelstudio/src/pages/Settings/StorageSettings/StorageSet.jsx

```diff
@@ -1,60 +1,73 @@
-import { useCallback, useContext, useEffect, useState } from "react";
+import { useCallback, useContext, useEffect, useMemo } from "react";
 import { Button, Columns } from "../../../components";
 import { confirm, modal } from "../../../components/Modal/Modal";
 import { Spinner } from "../../../components/Spinner/Spinner";
 import { ApiContext } from "../../../providers/ApiProvider";
-import { useProject } from "../../../providers/ProjectProvider";
+import { projectAtom } from "../../../providers/ProjectProvider";
 import { StorageCard } from "./StorageCard";
 import { StorageForm } from "./StorageForm";
+import { useAtomValue } from "jotai";
+import { useQuery } from "@tanstack/react-query";
 
 export const StorageSet = ({ title, target, rootClass, buttonLabel }) => {
   const api = useContext(ApiContext);
-  const { project } = useProject();
-  const [storages, setStorages] = useState([]);
-  const [loading, setLoading] = useState(false);
-  const [loaded, setLoaded] = useState(false);
-  const [storageTypes, setStorageTypes] = useState([]);
+  const project = useAtomValue(projectAtom);
+  const storageTypesQueryKey = ["storage-types", target];
+  const storagesQueryKey = ["storages", target, project?.id];
+
+  const {
```

Comment:
I would separate concerns here and put the API layer via React Query as hooks in the hooks folder